Move HandlerRegistry to RootHandler

tristan957 opened the pull request:
This was discussed over GroupMe, but handlers should only have
one regex expression. Each Handler then has sub Handlers that can do
further matching.

tristan957 committed:
Move HandlerRegistry to RootHandler

This was discussed over GroupMe, but handlers should only have
one regex expression. Each Handler then has sub Handlers that can do
further matching.

tristan957 commented:
conditional types are OP

JosephMart commented on src/handlers/Handler.ts:
```diff
@@ -2,7 +2,22 @@ import HandlerConfig from "./HandlerConfig";
+export default abstract class Handler<R = false> {
+  protected subHandlers: Handler[] = [];
+
+  readonly config: R extends false ? HandlerConfig : never;
```
What is this exactly?

tristan957 replied:
you talking about the type on config?

JosephMart replied:
R defaults to false and `config` extends `HandlerConfig` if only `true`?

tristan957 replied:
Comment needed, but essentially if you are the root handler you don't need a config is what I was going for.

If the root handler should have a config then I can amend it.

JosephMart commented on src/handlers/HeyHandler.ts:
```diff
@@ -7,15 +7,16 @@ import { Either } from "fp-ts/lib/Either";
-  async handle(groupMeInfo: GroupMeInfo): Promise<Either<Error, number>> {
-    return this.messagingService.sendMessage(`Howdy ${groupMeInfo.name}!`) as Promise<Either<Error, number>>;
+  handle(groupMeInfo: GroupMeInfo): Promise<Either<Error, unknown>>[] {
+    return super.handle(groupMeInfo).concat(this.messagingService.sendMessage(`Howdy ${groupMeInfo.name}!`));
```
Might want to add some comments. Kinda confused :P

tristan957 replied:
Ok :)

JosephMart commented on src/handlers/RootHandler.ts:
```diff
@@ -0,0 +1,28 @@
+    groupMeInfo = this.sanitizeText(groupMeInfo);
+    const matchedHandlers = this.subHandlers.filter(handler => this.shouldHandle(handler, groupMeInfo));
+
+    return matchedHandlers.map(handler => handler.handle(groupMeInfo)).reduce((prev, curr) => prev.concat(curr), []);
```
might want to `.filter` instead of `.reduce`? it is just appending to a running array?

tristan957 replied:
This takes a 2D array and transforms it into a 1D array

JosephMart replied:
There is a .flat or flatten func now in JS

JosephMart replied:
```suggestion
    return matchedHandlers.map(handler => handler.handle(groupMeInfo)).flat(Infinity);
```

tristan957 replied:
oh baby me make change 👍

JosephMart reviewed (approved):
LGTM with a few comments

tristan957 committed:
Use flat

Co-Authored-By: Joseph Martinsen <joseph@martinsen.com>

tristan957 committed:
Add documentation around the new structure of Handlers

tristan957 committed:
Make handler info-agnostic

Handler as an abstract class should be able to handle different types of information
depending on the service that it is implemented for.

JosephMart commented on api/loki.ts:
```diff
@@ -1,17 +1,17 @@
   container.register("MessagingService", { useValue: new GroupMeService() });
-  const handlerRegistry = container.resolve(HandlerRegistry);
-  const actions = handlerRegistry.delegate(groupMeInfo);
+  const handlerRegistry = container.resolve(RootHandler);
```
magic

JosephMart commented on src/handlers/HeyHandler.ts:
```diff
@@ -1,21 +1,35 @@
+   * @param groupMeInfo
+   */
+  shouldHandle = (groupMeInfo: GroupMeInfo): boolean => {
+    const passed = this.config.regexp.test(groupMeInfo.text);
+    return passed;
```
nit
```suggestion
    return this.config.regexp.test(groupMeInfo.text);
```

tristan957 replied:
I had it this way because conceivably there will be future checks like user who sent it, etc. Doesn't matter to me tho.

JosephMart commented on src/handlers/HeyHandler.ts:
```diff
@@ -1,21 +1,35 @@
+   * @param groupMeInfo
+   */
+  handle(groupMeInfo: GroupMeInfo): Promise<Either<Error, unknown>>[] {
+    const result = this.messagingService.sendMessage(`Howdy ${groupMeInfo.name}!`);
```
I think this needs to awaited
```suggestion
    const result = await this.messagingService.sendMessage(`Howdy ${groupMeInfo.name}!`);
```

tristan957 replied:
The await would be redundant. Since you don't resolve in the handle you let the caller resolve the handle which resolves the sendMessage.

JosephMart replied:
oh hm yeah guess so

JosephMart reviewed (approved):
besides the `await` suggestion, think LGTM

tristan957 committed:
Update src/handlers/HeyHandler.ts

Co-Authored-By: Joseph Martinsen <joseph@martinsen.com>
